cmake/Modules/FindPythonExtra.cmake

```diff
@@ -17,211 +17,75 @@
-      if(EXISTS "${_python_executable_debug}")
-        set(PYTHON_EXECUTABLE_DEBUG "${_python_executable_debug}")
-      else()
-        message(FATAL_ERROR "${_python_executable_debug} doesn't exist")
-      endif()
-    endif()
-    message(STATUS "Using PYTHON_EXECUTABLE: ${PYTHON_EXECUTABLE}")
-    message(STATUS "Using PYTHON_INCLUDE_DIRS: ${PYTHON_INCLUDE_DIRS}")
-    message(STATUS "Using PYTHON_LIBRARIES: ${PYTHON_LIBRARIES}")
-    set(PythonExtra_INCLUDE_DIRS "${PYTHON_INCLUDE_DIRS}")
-    set(PythonExtra_LIBRARIES "${PYTHON_LIBRARIES}")
-  endif()
-
-  if(NOT DEFINED PYTHON_SOABI)
-    set(_python_code
-      "from sysconfig import get_config_var"
-      "print(get_config_var('SOABI'))"
-    )
-    execute_process(
-      COMMAND
-      "${PYTHON_EXECUTABLE}"
-      "-c"
-      "${_python_code}"
-      OUTPUT_VARIABLE _output
-      RESULT_VARIABLE _result
-      OUTPUT_STRIP_TRAILING_WHITESPACE
-    )
-    if(NOT _result EQUAL 0)
-      message(FATAL_ERROR
-        "execute_process(${PYTHON_EXECUTABLE} -c '${_python_code}') returned "
-        "error code ${_result}")
-    endif()
-
-    set(PYTHON_SOABI
-      "${_output}"
-      CACHE INTERNAL
-      "The SOABI suffix for Python native extensions. See PEP-3149: https://www.python.org/dev/peps/pep-3149/.")
-  endif()
-
-  if(PYTHON_SOABI STREQUAL "" OR PYTHON_SOABI STREQUAL "None")
-    set(PythonExtra_EXTENSION_SUFFIX
-      ""
-      CACHE INTERNAL
-      "The full suffix for Python native extensions. See PEP-3149: https://www.python.org/dev/peps/pep-3149/."
-    )
-  else()
-    set(PythonExtra_EXTENSION_SUFFIX
-      ".${PYTHON_SOABI}"
-      CACHE INTERNAL
-      "The full suffix for Python native extensions. See PEP-3149: https://www.python.org/dev/peps/pep-3149/."
-    )
-  endif()
+if(PythonExtra_FOUND)
+  return()
+endif()
 
-  if(WIN32)
-    if(CMAKE_BUILD_TYPE STREQUAL "Debug")
-      set(PythonExtra_EXTENSION_EXTENSION "_d.pyd")
-    else()
-      set(PythonExtra_EXTENSION_EXTENSION ".pyd")
-    endif()
-  else()
-    # Also use .so for OSX, not dylib
-    set(PythonExtra_EXTENSION_EXTENSION ".so")
+find_package(Python3 REQUIRED COMPONENTS Interpreter)
+
+get_target_property(PYTHON_EXECUTABLE Python3::Interpreter LOCATION)
+
+add_executable(PythonExtra::Interpreter IMPORTED)
+set_property(TARGET PythonExtra::Interpreter
+  PROPERTY IMPORTED_LOCATION "${PYTHON_EXECUTABLE}")
+
+# Set the location to the debug interpreter on Windows if it exists
+if(WIN32)
+  get_filename_component(_python_executable_dir "${PYTHON_EXECUTABLE}" DIRECTORY)
+  get_filename_component(_python_executable_name "${PYTHON_EXECUTABLE}" NAME_WE)
+  get_filename_component(_python_executable_ext "${PYTHON_EXECUTABLE}" EXT)
+  set(PYTHON_EXECUTABLE_DEBUG "${_python_executable_dir}/${_python_executable_name}_d${_python_executable_ext}")
+  if(EXISTS "${PYTHON_EXECUTABLE_DEBUG}")
+    set_property(TARGET PythonExtra::Interpreter PROPERTY IMPORTED_LOCATION "${PYTHON_EXECUTABLE_DEBUG}")
```

Comment:
It looks like this sets `PythonExtra::Interpreter` to the debug version regardless of the build type. 

How about we set `IMPORTED_LOCATION_Debug` instead? It looks like systems with both a release and a debug version of Python would use the debug version only when the build type was set to `Debug` if I understand these correctly 
* https://cmake.org/cmake/help/latest/prop_tgt/IMPORTED_LOCATION_CONFIG.html
* https://cmake.org/cmake/help/latest/prop_tgt/IMPORTED_LOCATION.html#prop_tgt:IMPORTED_LOCATION

Reply:
This is something I've never understood about this whole system.  Don't we want to unconditionally set the python interpreter to the debug variant in a debug build?  Why do we set it as this side variable, and then have to sprinkle `if(WIN32 && DEBUG)` throughout all of the downstream CMakeLists.txt?  Shouldn't we just set it once and then use that everywhere?

Reply:
>   Don't we want to unconditionally set the python interpreter to the debug variant in a debug build? Why do we set it as this side variable, and then have to sprinkle if(WIN32 && DEBUG) throughout all of the downstream CMakeLists.txt? Shouldn't we just set it once and then use that everywhere?

IIUC `IMPORTED_LOCATION_Debug` overrides `IMPORTED_LOCATION` when the build type is set to `Debug`,  so we no longer need `if(..)` code anywhere because CMake would do that for us.

Reply:
OK, after doing some debugging on this, including trying out the basic examples from https://cmake.org/cmake/help/latest/guide/importing-exporting/index.html#importing-libraries , I just can't make this work.  Either this is buggy in CMake, or there is something we don't understand about how this works.

So for now, we are always overriding `IMPORTED_LOCATION` with the appropriate non-debug/debug interpreter, and this seems to work.

In addition, we decided to go with a new variable exported from python_cmake_module called `PythonExtra_POSTFIX`.  This should be set by downstream consumers on their targets using `set_target_properties(${_target_name} PROPERTIES DEBUG_POSTFIX "${PythonExtra_POSTFIX}")`, which will make them always use the _d libraries as necessary.

With that, I think these are good to go.  I'm going to run another round of CI here to see what happens.